File: js/src/accounts/Store.ts
import {
  Borsh,
  StringPublicKey,
  ERROR_INVALID_OWNER,
  AnyPublicKey,
  Account,
} from '@metaplex-foundation/mpl-core';
import { AccountInfo } from '@solana/web3.js';
import BN from 'bn.js';
import { AccountKey } from './constants';
import { PassBookProgram } from '../PassBookProgram';

type Args = {
  key: AccountKey;
  authority: StringPublicKey;
  redemptionsCount: BN;
  passCount: BN;
  passBookCount: BN;
};

export class StoreData extends Borsh.Data<Args> {
  static readonly SCHEMA = StoreData.struct([
    ['key', 'u8'],
    ['authority', 'pubkeyAsString'],
    ['redemptionsCount', 'u64'],
    ['passCount', 'u64'],
    ['passBookCount', 'u64'],
  ]);
  key: AccountKey;
  mint: StringPublicKey;
  authority: StringPublicKey;
  redemptionsCount: string;
  passCount: string;
  passBookCount: string;

  constructor(args: Args) {
    super(args);
    this.key = AccountKey.PassStore;
  }
}

export class Store extends Account<StoreData> {
  constructor(pubkey: AnyPublicKey, info: AccountInfo<Buffer>) {
    super(pubkey, info);
    if (!this.assertOwner(PassBookProgram.PUBKEY)) {
      throw ERROR_INVALID_OWNER();
    }
    this.data = StoreData.deserialize(this.info.data);
  }
}

File: dart/lib/accounts/store.dart
import 'dart:typed_data';

import 'package:passbook/accounts/constants.dart';
import 'package:passbook/utils/endian.dart';
import 'package:passbook/utils/struct_reader.dart';
import 'package:solana/base58.dart';

class Store {
  const Store({
    required this.key,
    required this.authority,
    required this.redemptionsCount,
    required this.passCount,
    required this.passBookCount,
  });

  static const prefix = 'store';

  factory Store.fromBinary(List<int> sourceBytes) {
    final bytes = Int8List.fromList(sourceBytes);
    final reader = StructReader(bytes.buffer)..skip(1);
    final authority = base58encode(reader.nextBytes(32));
    final redemptionsCount = decodeBigInt(reader.nextBytes(8), Endian.little);
    final passCount = decodeBigInt(reader.nextBytes(8), Endian.little);
    final passBookCount = decodeBigInt(reader.nextBytes(8), Endian.little);
    return Store(
        key: AccountKey.passStore,
        authority: authority,
        redemptionsCount: redemptionsCount,
        passCount: passCount,
        passBookCount: passBookCount);
  }

  final AccountKey key;
  final String authority;
  final BigInt redemptionsCount;
  final BigInt passCount;
  final BigInt passBookCount;
}
